plutus-use-cases/src/Language/Plutus/Coordination/Contracts/CrowdFunding.hs

```diff
@@ -0,0 +1,150 @@
+
+import           Prelude                              (Bool (..), Maybe (..), Num (..), Ord (..), succ, sum, ($))
+
+-- | A crowdfunding campaign.
+data Campaign = Campaign
+    { campaignDeadline           :: !BlockHeight
+    , campaignTarget             :: !Value
+    , campaignCollectionDeadline :: !BlockHeight
+    , campaignOwner              :: !PubKey
+    }
+
+-- | Contribute funds to the campaign (contributor)
+--
+contribute :: Campaign -> Value -> TxM [TxOutRef]
+contribute c value = do
+    assert (value > 0)
+    contributorPubKey <- lookupMyPubKey
+    myPayment         <- createPayment (value + standardTxFee)
+    let validator = contributionScript c (Just contributorPubKey)
+        o = TxOutScript
+            value
+            (hash validator)
+            0 -- TODO: contributorPubKey ought to be lifted into PLC at coordination runtime as the data script
+    submitTransaction Tx
+      { txInputs  = (myPayment, myPayment) -- TODO: Change to [myPayment] when we can have a list of inputs
+      , txOutputs = (o, o)
+      }
+    -- the transaction above really ought to be merely a transaction *template* and the transaction fee ought to be
+    -- added by the Wallet API Plutus library on the basis of the size and other costs of the transaction
+
+-- | The validator script that determines whether the campaign owner can
+--   retrieve the funds or the contributors can claim a refund.
+--
+--   See note [Contracts and Validator Scripts] in
+--       Language.Plutus.Coordination.Contracts
+contributionScript ::
+       Campaign
+    -> Maybe PubKey
+    -> PlutusTx
+contributionScript _ _  = PlutusTx inner where
+    inner = plc  (\() () p Campaign{..} contribPubKey ->
+        let
+            -- | Check that a transaction input is signed by the private key of the given
+            --   public key.
+            signedBy :: TxIn -> PubKey -> Bool
+            signedBy = Prim.error
+
+            infixr 3 &&
+            (&&) :: Bool -> Bool -> Bool
+            (&&) = Prim.error
+
+            infixr 3 ||
+            (||) :: Bool -> Bool -> Bool
+            (||) = Prim.error
+
+            -- | Check that a pending transcation is signed by the private key of the given
+            --   public key.
+            signedByT :: PendingTx -> PubKey -> Bool
+            signedByT = Prim.error
+
+            maybe :: b -> (a -> b) -> Maybe a -> b
+            maybe = Prim.error
+
+            PendingTx pendingTxBlockHeight _ pendingTxTransaction = p
+
+            Tx (t1, t2) _ = pendingTxTransaction
+            TxIn (TxOutRef v1 _ _) _ = t1
+            TxIn (TxOutRef v2 _ _) _ = t2
+
+
+            pledgedFunds = v1 + v2
+            -- Check that a refund transaction only spends the amount that was
+            -- pledged by the contributor identified by `contribPubKey`
+            contributorOnly = maybe False (\k ->
+                                signedBy t1 k && signedBy t2 k) contribPubKey
```

Comment:
This looks like it's checking that both inputs were from the same contributor, which doesn't seem right.

Reply:
Yeah the problem here is that the txn input size (as defined in Language.Plutus.Coordination.Plutus.Tx) is now fixed to two, but ideally in the "refund" case we would only have one txn. I think there is some inconsistency in the model at the moment :|